spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/AbstractConnectionFactory.java

```diff
@@ -595,8 +595,10 @@ else if (soTimeout > 0) {
 								logger.warn("Timing out TcpNioConnection " +
 										    connection.getConnectionId());
 							}
-							connection.publishConnectionExceptionEvent(new SocketTimeoutException("Timing out connection"));
+							SocketTimeoutException exception = new SocketTimeoutException("Timing out connection");
+							connection.publishConnectionExceptionEvent(exception);
 							connection.timeout();
+							connection.sendExceptionToListener(exception);
```

Comment:
No, thanks, Gary.
I'm fine with the explanation.
My confuse was about the remote connect, like how can we send anything to the listener if we have just closed the socket...
Never mind, some my brain mess. We have everything locally here. And we really just have an issue around `soTimeout`, which just has been ignored before.

So, merging ...